xcm/pallet-xcm/src/lib.rs

```diff
@@ -1098,6 +1099,67 @@ pub mod pallet {
 /// The maximum number of distinct assets allowed to be transferred in a single helper extrinsic.
 const MAX_ASSETS_FOR_TRANSFER: usize = 2;
 
+impl<T: Config> XcmQueryHandler for Pallet<T> {
+	type QueryId = u64;
+	type BlockNumber = T::BlockNumber;
+	type Error = XcmError;
+
+	/// To check the status of the query, use `fn query()` passing the resultant `QueryId`
+	/// value.
+	fn report_outcome(
+		message: &mut Xcm<()>,
+		responder: impl Into<MultiLocation>,
+		timeout: Self::BlockNumber,
+	) -> Result<Self::QueryId, Self::Error> {
+		let responder = responder.into();
+		let destination = T::UniversalLocation::get()
+			.invert_target(&responder)
+			.map_err(|()| XcmError::LocationNotInvertible)?;
+		let query_id = Self::new_query(responder, timeout, Here);
+		let response_info = QueryResponseInfo { destination, query_id, max_weight: Weight::zero() };
+		let report_error = Xcm(vec![ReportError(response_info)]);
+		message.0.insert(0, SetAppendix(report_error));
+		Ok(query_id)
+	}
+
+	fn take_response(query_id: Self::QueryId) -> QueryResponseStatus<Self::BlockNumber> {
+		match Queries::<T>::get(query_id) {
+			Some(QueryStatus::Ready { response, at }) => {
+				let response = response.try_into();
+				match response {
+					Ok(response) => {
+						Queries::<T>::remove(query_id);
+						Self::deposit_event(Event::ResponseTaken(query_id));
+						QueryResponseStatus::Finished(FinishedQuery::Response { response, at })
+					},
+					Err(_) => QueryResponseStatus::UnexpectedVersion,
+				}
+			},
+			Some(QueryStatus::VersionNotifier { origin, is_active }) => {
```

Comment:
I was wrong with `VersionNotifier`, it's not relevant in this context. `VersionNotifier` only occurs for the queries from XCM version subscription. I would just explicitly map it to NotFound and leave the comment.